torch/csrc/jit/passes/peephole_non_tensor.cpp

```diff
@@ -114,6 +114,26 @@ struct PeepholeOptimizeNonTensorImpl {
           node->output()->replaceAllUsesWith(node->input());
           changed = true;
         }
+      } else if (node->kind() == aten::ne || node->kind() == aten::eq) {
+        if (node->inputs().size() != 2 ||
+            node->inputs().at(0) != node->inputs().at(1)) {
+          continue;
+        }
+        auto inp_kind = node->inputs().at(0)->type()->kind();
+        // only handling common immutable types here because other types like
+        // Tensor or list of Tensor might throw on aten::eq
+        switch (inp_kind) {
+          case TypeKind::BoolType:
+          case TypeKind::IntType:
+          case TypeKind::FloatType: {
```

Comment:
What about `ComplexType` and `StringType`?

BTW, in which case would `x == x` throw exceptions on mutable types?

Reply:
Tensor and anything containing Tensor